pkg/controllers/management/auth/project_cluster/common_test.go

```diff
@@ -0,0 +1,174 @@
+						Name: "test-project-owner",
+						OwnerReferences: []metav1.OwnerReference{
+							{
+								APIVersion: "v3",
+								Kind:       "Project",
+								Name:       "test-project",
+								UID:        "1234abcd",
+							},
+						},
+					},
+					Rules: []rbacv1.PolicyRule{
+						{
+							APIGroups:     []string{"management.cattle.io"},
+							Resources:     []string{"projects"},
+							ResourceNames: []string{"test-project"},
+							Verbs:         []string{"*"},
+						},
+					},
+				},
+			},
+		},
+		{
+			name: "roles for cluster",
+			obj: &apisv3.Cluster{
+				TypeMeta: metav1.TypeMeta{
+					APIVersion: "v3",
+					Kind:       "Cluster",
+				},
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "test-cluster",
+					UID:  "1234abcd",
+				},
+			},
+			wantedCRs: []*rbacv1.ClusterRole{
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: "test-cluster-member",
+						OwnerReferences: []metav1.OwnerReference{
+							{
+								APIVersion: "v3",
+								Kind:       "Cluster",
+								Name:       "test-cluster",
+								UID:        "1234abcd",
+							},
+						},
+						Annotations: map[string]string{
+							"cluster.cattle.io/name": "test-cluster",
+						},
+					},
+					Rules: []rbacv1.PolicyRule{
+						{
+							APIGroups:     []string{"management.cattle.io"},
+							Resources:     []string{"clusters"},
+							ResourceNames: []string{"test-cluster"},
+							Verbs:         []string{"get"},
+						},
+					},
+				},
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: "test-cluster-owner",
+						OwnerReferences: []metav1.OwnerReference{
+							{
+								APIVersion: "v3",
+								Kind:       "Cluster",
+								Name:       "test-cluster",
+								UID:        "1234abcd",
+							},
+						},
+						Annotations: map[string]string{
+							"cluster.cattle.io/name": "test-cluster",
+						},
+					},
+					Rules: []rbacv1.PolicyRule{
+						{
+							APIGroups:     []string{"management.cattle.io"},
+							Resources:     []string{"clusters"},
+							ResourceNames: []string{"test-cluster"},
+							Verbs:         []string{"*"},
+						},
+					},
+				},
+			},
+		},
+		{
+			name:    "called with unsupported type",
+			obj:     &apisv3.ProjectNetworkPolicyList{},
+			wantErr: true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			ctrl := gomock.NewController(t)
+			crClient := fake.NewMockNonNamespacedControllerInterface[*rbacv1.ClusterRole, *rbacv1.ClusterRoleList](ctrl)
+
+			if tt.wantedCRs != nil {
+				crClient.EXPECT().Get(gomock.Any(), gomock.Any()).Return(nil, errNotFound).AnyTimes()
+				crClient.EXPECT().Create(gomock.Any()).DoAndReturn(func(c *rbacv1.ClusterRole) (*rbacv1.ClusterRole, error) {
+					assert.Equal(t, tt.wantedCRs[0], c)
+					return nil, nil
+				})
+				crClient.EXPECT().Create(gomock.Any()).DoAndReturn(func(c *rbacv1.ClusterRole) (*rbacv1.ClusterRole, error) {
+					assert.Equal(t, tt.wantedCRs[1], c)
+					return nil, nil
+				})
```

Comment:
We should avoid gomock.Any() and AnyTimes() if possible to have more accurate tests. This way we also verify the parameters passed and the times the mock methods are called

```suggestion
				crClient.EXPECT().Get(tt.wantedCRs[0].Name, metav1.GetOptions{}).Return(nil, errNotFound)
				crClient.EXPECT().Get(tt.wantedCRs[1].Name, metav1.GetOptions{}).Return(nil, errNotFound)
				crClient.EXPECT().Create(tt.wantedCRs[0])
				crClient.EXPECT().Create(tt.wantedCRs[1])
```

Reply:
Fixed in b88c01d892996f7f2b83622717cd927020e92387